### PR TITLE
AWS: Enable RetryMode for AWS KMS client

### PR DESCRIPTION
**Pull request by hsiang-c**
* We're reaching the limit of AWS KMS requests per AWS account quota and we'd like to enable retry in KMS client to be more resilient to failures.
* This PR is similar to https://github.com/apache/iceberg/pull/11052 but using the SDK provided `RetryMode`.
* I didn't choose `RetryPolicy` b/c it is deprecated.

### CODE REVIEW

**Comment by hsiang-c** [comment]
@nastra @jackye1995 please review it when you have time, thanks!

File: aws/src/main/java/org/apache/iceberg/aws/kms/KmsClientProperties.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  public static final String KMS_RETRY_MODE_DEFAULT = "LEGACY";
+
+  private RetryMode kmsRetryMode;
+
+  public KmsClientProperties() {
+    this.kmsRetryMode = RetryMode.valueOf(KMS_RETRY_MODE_DEFAULT);
+  }
+
+  public KmsClientProperties(Map<String, String> properties) {
```
Comment (jbonofre) [inline aws/src/main/java/org/apache/iceberg/aws/kms/KmsClientProperties.java:41]: nit: I would add the link to AWS configBuilder configuration, especially about retry mode (to know the accepted values for retry mode).

**Review by jbonofre — APPROVED** [review]
It looks good to me. I just left a nit about pointing to the documentation (especially about retry mode accepted values).

**Comment by danielcweeks** [comment]
Is this really necessary to make configurable?  Why shouldn't we just set `ADAPTIVE_V2` as the default when constructing the client?  This is only for retry behavior, I'm not sure it's worth even making it an option.

**Comment by hsiang-c** [comment]
@danielcweeks Thank you for your review.

> Is this really necessary to make configurable?

I am also working on an enhancement to Iceberg S3 doc: https://github.com/apache/iceberg/pull/11321, which makes me think users might want to configure AWS service clients with available configurations provided by the client libraries.

> This is only for retry behavior, I'm not sure it's worth even making it an option.

Got it, I have no problem with setting `ADAPTIVE_V2` as the default, though the [default mode](https://github.com/aws/aws-sdk-java-v2/blob/2.29.7/core/sdk-core/src/main/java/software/amazon/awssdk/core/retry/RetryMode.java#L133) is `LEGACY` according to AWS Java V2 SDK.

Is having too many options a concern to you? I'd like to learn more about your thoughts.

Thanks again for your time.

**Comment by danielcweeks** [comment]
> Is having too many options a concern to you? I'd like to learn more about your thoughts.

I feel like the proliferation of options is not a good thing unless there's real value being added.  This doesn't seem to be one of those cases.  Rather than adding a lot of boilerplate configuration, I would suggest just setting this at `ADAPTIVE_V2`, which seems to be an improvement upon the legacy default.

**Comment by hsiang-c** [comment]
Thanks for your feedback @danielcweeks Please take a look again.

If you still have time, I have another PR for you https://github.com/apache/iceberg/pull/11321

**Review by amogh-jahagirdar — COMMENTED** [review]
Thanks @hsiang-c I agree with @danielcweeks that we need to be cautious when adding new properties, specifically what that means is evaluating if it's a valuable knob that's worth adding. In this case we know that we generally want to retry when throttled and the recommended strategy is adaptive so the produced clients should just do that without exposing things to users. 

Overall this PR looks to be a good direction, just had a comment that I don't think the public KmsClientProperties class is needed at the moment.

File: aws/src/main/java/org/apache/iceberg/aws/kms/KmsClientProperties.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * under the License.
+ */
+package org.apache.iceberg.aws.kms;
+
+import java.io.Serializable;
+import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration;
+import software.amazon.awssdk.core.retry.RetryMode;
+import software.amazon.awssdk.services.kms.KmsClientBuilder;
+
+public class KmsClientProperties implements Serializable {
```
[the remark this answers is not in the document]
Comment (hsiang-c) [reply aws/src/main/java/org/apache/iceberg/aws/kms/KmsClientProperties.java:26]: @amogh-jahagirdar Thank you for your review.

Making `applyRetryConfigurations ` as part of `AwsClientProperties` makes sense to me.

File: aws/src/test/java/org/apache/iceberg/aws/kms/TestKmsClientProperties.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+public class TestKmsClientProperties {
+  @Test
+  public void testApplyRetryConfiguration() {
+    AwsClientProperties awsClientProperties = new AwsClientProperties();
+
+    KmsClientBuilder builder = KmsClient.builder();
+    awsClientProperties.applyRetryConfigurations(builder);
+    RetryMode retryPolicy = builder.overrideConfiguration().retryMode().get();
+
+    assertThat(retryPolicy).as("retry mode was not set").isEqualTo(RetryMode.ADAPTIVE_V2);
```
Comment (amogh-jahagirdar) [inline aws/src/test/java/org/apache/iceberg/aws/kms/TestKmsClientProperties.java:38]: Nit: Could we clean up this error message to be "Retry mode should be ADAPTIVE_V2"?

**Review by amogh-jahagirdar — DISMISSED** [review]

File: aws/src/main/java/org/apache/iceberg/aws/AwsClientProperties.java [hunk 4]
```diff
@@ -154,6 +156,23 @@ public AwsCredentialsProvider credentialsProvider(
...
   }
 
+  /**
+   * Set ADAPTIVE_V2 <a
+   * href="https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/retry/RetryMode.html">RetryMode</a>
+   * for an AWS client builder.
+   *
+   * <p>For more details, see
+   * https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.Builder.html#retryStrategy(software.amazon.awssdk.core.retry.RetryMode)
+   */
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/AwsClientProperties.java:166]: Could we structure this JavaDoc similar to the  the other apply configuration methods? e.g. checkout the java doc in applyClientCredentialConfigurations or applyClientRegionConfiguration

**Review by amogh-jahagirdar — COMMENTED** [review]
Thanks @hsiang-c I think this is very close just some java docs that I think should be cleaned up before merging

File: aws/src/test/java/org/apache/iceberg/aws/kms/TestKmsClientProperties.java (hunk 3 above)
Comment (hsiang-c) [reply aws/src/test/java/org/apache/iceberg/aws/kms/TestKmsClientProperties.java:38]: Sorry I missed this one.

**Review by amogh-jahagirdar — APPROVED** [review]
Thanks @hsiang-c , this looks great to me now!

**Comment by amogh-jahagirdar** [comment]
I'll go ahead and merge

**Comment by hsiang-c** [comment]
Thank you @amogh-jahagirdar

**Pull request merged**